crates/red_knot_project/src/metadata/value.rs

```diff
@@ -59,115 +71,276 @@ impl Drop for ValueSourceGuard {
+#[derive(Clone)]
+pub struct RangedValue<T> {
+    value: T,
     source: ValueSource,
+
+    /// The byte range of `value` in `source`.
+    ///
+    /// Can be `None` because not all sources support a range.
+    /// For example, arguments provided on the CLI won't have a range attached.
+    range: Option<TextRange>,
 }
 
-impl RelativePathBuf {
-    pub fn new(path: impl AsRef<SystemPath>, source: ValueSource) -> Self {
+impl<T> RangedValue<T> {
+    pub fn new(value: T, source: ValueSource) -> Self {
+        Self::with_range(value, source, TextRange::default())
+    }
+
+    pub fn cli(value: T) -> Self {
+        Self::with_range(value, ValueSource::Cli, TextRange::default())
+    }
+
+    pub fn with_range(value: T, source: ValueSource, range: TextRange) -> Self {
         Self {
-            path: path.as_ref().to_path_buf(),
+            value,
+            range: Some(range),
             source,
         }
     }
 
-    pub fn cli(path: impl AsRef<SystemPath>) -> Self {
-        Self::new(path, ValueSource::Cli)
+    pub fn range(&self) -> Option<TextRange> {
+        self.range
     }
 
-    /// Returns the relative path as specified by the user.
-    pub fn path(&self) -> &SystemPath {
-        &self.path
+    pub fn source(&self) -> &ValueSource {
+        &self.source
     }
 
-    /// Returns the owned relative path.
-    pub fn into_path_buf(self) -> SystemPathBuf {
-        self.path
+    #[must_use]
+    pub fn with_source(mut self, source: ValueSource) -> Self {
+        self.source = source;
+        self
     }
 
-    /// Resolves the absolute path for `self` based on from where the value origins.
-    pub fn absolute_with_db(&self, db: &dyn Db) -> SystemPathBuf {
-        self.absolute(db.project().root(db), db.system())
+    pub fn into_inner(self) -> T {
+        self.value
     }
+}
 
-    /// Resolves the absolute path for `self` based on from where the value origins.
-    pub fn absolute(&self, project_root: &SystemPath, system: &dyn System) -> SystemPathBuf {
-        let relative_to = match &self.source {
-            ValueSource::File(_) => project_root,
-            ValueSource::Cli => system.current_directory(),
-        };
+impl<T> Combine for RangedValue<T> {
+    fn combine(self, _other: Self) -> Self
+    where
+        Self: Sized,
+    {
+        self
+    }
+    fn combine_with(&mut self, _other: Self) {}
+}
 
-        SystemPath::absolute(&self.path, relative_to)
+impl<T> IntoIterator for RangedValue<T>
+where
+    T: IntoIterator,
+{
+    type Item = T::Item;
+    type IntoIter = T::IntoIter;
+    fn into_iter(self) -> Self::IntoIter {
+        self.value.into_iter()
     }
 }
 
-// TODO(micha): Derive most of those implementations once `RelativePath` uses `Value`.
-//   and use `serde(transparent, deny_unknown_fields)`
-impl Combine for RelativePathBuf {
-    fn combine(self, _other: Self) -> Self {
-        self
+// The type already has a `into_iter` method thanks to `Deref`.
```

Comment:
This comment doesn't make sense to me, because we are defining the `into_iter` method for the `RangedValue` type here. Did you mean to say that it already has an `iter` method?